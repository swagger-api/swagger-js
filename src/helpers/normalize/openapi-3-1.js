import { dispatchRefractorPlugins, isObjectElement } from '@swagger-api/apidom-core';
import {
  refractorPluginNormalizeOperationIds,
  refractorPluginNormalizeParameters,
  refractorPluginNormalizeSecurityRequirements,
  refractorPluginNormalizeServers,
  refractorPluginNormalizeParameterExamples,
  refractorPluginNormalizeHeaderExamples,
  createToolbox,
  keyMap,
  getNodeType,
} from '@swagger-api/apidom-ns-openapi-3-1';

import opId from '../op-id.js';

const normalize = (element) => {
  if (!isObjectElement(element)) return element;
  if (element.hasKey('$$normalized')) return element;

  const plugins = [
    refractorPluginNormalizeOperationIds({
      operationIdNormalizer: (operationId, path, method) =>
        opId({ operationId }, path, method, { v2OperationIdCompatibilityMode: false }),
    }),
    refractorPluginNormalizeParameters(),
    refractorPluginNormalizeSecurityRequirements(),
    refractorPluginNormalizeServers(),
    refractorPluginNormalizeParameterExamples(),
    refractorPluginNormalizeHeaderExamples(),
  ];

  const normalized = dispatchRefractorPlugins(element, plugins, {
    toolboxCreator: createToolbox,
    visitorOptions: { keyMap, nodeTypeGetter: getNodeType },
  });

  normalized.set('$$normalized', true);
  return normalized;
};

export default normalize;
