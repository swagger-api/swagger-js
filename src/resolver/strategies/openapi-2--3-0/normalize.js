import opId from '../../../helpers/op-id.js';

export default function normalize(parsedSpec) {
  const { spec } = parsedSpec;
  const { paths } = spec;
  const map = {};

  if (!paths || spec.$$normalized) {
    return parsedSpec;
  }

  // eslint-disable-next-line no-restricted-syntax, guard-for-in
  for (const pathName in paths) {
    const path = paths[pathName];

    if (path == null || !['object', 'function'].includes(typeof path)) {
      continue; // eslint-disable-line no-continue
    }

    const pathParameters = path.parameters;

    // eslint-disable-next-line no-restricted-syntax, guard-for-in
    for (const method in path) {
      const operation = path[method];
      if (operation == null || !['object', 'function'].includes(typeof operation)) {
        continue; // eslint-disable-line no-continue
      }

      const oid = opId(operation, pathName, method);

      if (oid) {
        if (map[oid]) {
          map[oid].push(operation);
        } else {
          map[oid] = [operation];
        }

        const opList = map[oid];
        if (opList.length > 1) {
          opList.forEach((o, i) => {
            // eslint-disable-next-line no-underscore-dangle
            o.__originalOperationId = o.__originalOperationId || o.operationId;
            o.operationId = `${oid}${i + 1}`;
          });
        } else if (typeof operation.operationId !== 'undefined') {
          // Ensure we always add the normalized operation ID if one already exists
          // ( potentially different, given that we normalize our IDs)
          // ... _back_ to the spec. Otherwise, they might not line up
          const obj = opList[0];
          // eslint-disable-next-line no-underscore-dangle
          obj.__originalOperationId = obj.__originalOperationId || operation.operationId;
          obj.operationId = oid;
        }
      }

      if (method !== 'parameters') {
        // Add inherited consumes, produces, parameters, securities
        const inheritsList = [];
        const toBeInherit = {};

        // Global-levels
        // eslint-disable-next-line no-restricted-syntax
        for (const key in spec) {
          if (key === 'produces' || key === 'consumes' || key === 'security') {
            toBeInherit[key] = spec[key];
            inheritsList.push(toBeInherit);
          }
        }

        // Path-levels
        if (pathParameters) {
          toBeInherit.parameters = pathParameters;
          inheritsList.push(toBeInherit);
        }

        if (inheritsList.length) {
          // eslint-disable-next-line no-restricted-syntax
          for (const inherits of inheritsList) {
            // eslint-disable-next-line no-restricted-syntax
            for (const inheritName in inherits) {
              if (!operation[inheritName]) {
                operation[inheritName] = inherits[inheritName];
              } else if (inheritName === 'parameters') {
                // eslint-disable-next-line no-restricted-syntax
                for (const param of inherits[inheritName]) {
                  const exists = operation[inheritName].some(
                    (opParam) =>
                      (opParam.name && opParam.name === param.name) ||
                      (opParam.$ref && opParam.$ref === param.$ref) ||
                      (opParam.$$ref && opParam.$$ref === param.$$ref) ||
                      opParam === param
                  );

                  if (!exists) {
                    operation[inheritName].push(param);
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  spec.$$normalized = true;

  return parsedSpec;
}
