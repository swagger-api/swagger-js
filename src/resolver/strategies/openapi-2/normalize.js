// eslint-disable-next-line no-restricted-exports
export { default } from '../generic/normalize.js';
