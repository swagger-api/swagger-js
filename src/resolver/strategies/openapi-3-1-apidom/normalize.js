/* eslint-disable camelcase */
import { dispatchRefractorPlugins, isObjectElement, toValue } from '@swagger-api/apidom-core';
import {
  refractorPluginNormalizeOperationIds,
  refractorPluginNormalizeParameters,
  refractorPluginNormalizeSecurityRequirements,
  refractorPluginNormalizeServers,
  refractorPluginNormalizeParameterExamples,
  refractorPluginNormalizeHeaderExamples,
  createToolbox,
  keyMap,
  getNodeType,
  OpenApi3_1Element,
} from '@swagger-api/apidom-ns-openapi-3-1';

import opId from '../../../helpers/op-id.js';

const normalize = (element) => {
  if (!isObjectElement(element)) return element;
  if (element.hasKey('$$normalized')) return element;

  const plugins = [
    refractorPluginNormalizeOperationIds({
      operationIdNormalizer: (operationId, path, method) =>
        opId({ operationId }, path, method, { v2OperationIdCompatibilityMode: false }),
    }),
    refractorPluginNormalizeParameters(),
    refractorPluginNormalizeSecurityRequirements(),
    refractorPluginNormalizeServers(),
    refractorPluginNormalizeParameterExamples(),
    refractorPluginNormalizeHeaderExamples(),
  ];

  const normalized = dispatchRefractorPlugins(element, plugins, {
    toolboxCreator: createToolbox,
    visitorOptions: { keyMap, nodeTypeGetter: getNodeType },
  });

  normalized.set('$$normalized', true);
  return normalized;
};

/**
 * This adapter allow to perform normalization on Plain Old JavaScript Objects.
 * The function adapts the `normalize` function interface and is able to accept
 * Plain Old JavaScript Objects and returns Plain Old JavaScript Objects.
 */
export const pojoAdapter = (normalizeFn) => (spec) => {
  if (spec?.$$normalized) return spec;
  if (pojoAdapter.cache.has(spec)) return spec;

  const openApiElement = OpenApi3_1Element.refract(spec);
  const normalized = normalizeFn(openApiElement);
  const value = toValue(normalized);

  pojoAdapter.cache.set(spec, value);

  return value;
};
pojoAdapter.cache = new WeakMap();

export default normalize;
/* eslint-enable camelcase */
