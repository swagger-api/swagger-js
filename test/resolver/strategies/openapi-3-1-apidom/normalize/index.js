import path from 'node:path';
/* eslint-disable camelcase */
import { toValue, StringElement } from '@swagger-api/apidom-core';
import { OpenApi3_1Element } from '@swagger-api/apidom-ns-openapi-3-1';

import normalize from '../../../../../src/resolver/strategies/openapi-3-1-apidom/normalize.js';

const fixturesPath = path.join(__dirname, '__fixtures__');

/**
 * This test suite only guarantees that normalization works for OpenAPI 3.1.0.
 * Complete coverage of all use-cases including corner-cases are covered
 * upstream in ApiDOM.
 */

describe('helpers', () => {
  describe('normalize', () => {
    describe('OpenAPI 3.1', () => {
      describe('given denormalized Header Object examples', () => {
        test('should normalize', () => {
          const spec = globalThis.loadJsonFile(path.join(fixturesPath, 'header-examples.json'));
          const openApiElement = OpenApi3_1Element.refract(spec);
          const normalized = normalize(openApiElement);

          expect(toValue(normalized)).toMatchSnapshot();
        });
      });

      describe('given denormalized Parameter Object examples', () => {
        test('should normalize', () => {
          const spec = globalThis.loadJsonFile(path.join(fixturesPath, 'parameter-examples.json'));
          const openApiElement = OpenApi3_1Element.refract(spec);
          const normalized = normalize(openApiElement);

          expect(toValue(normalized)).toMatchSnapshot();
        });
      });

      describe('given denormalized Operation.id fields', () => {
        test('should normalize', () => {
          const spec = globalThis.loadJsonFile(path.join(fixturesPath, 'operation-ids.json'));
          const openApiElement = OpenApi3_1Element.refract(spec);
          const normalized = normalize(openApiElement);

          expect(toValue(normalized)).toMatchSnapshot();
        });
      });

      describe('given denormalized Parameter Objects', () => {
        test('should normalize', () => {
          const spec = globalThis.loadJsonFile(path.join(fixturesPath, 'parameters.json'));
          const openApiElement = OpenApi3_1Element.refract(spec);
          const normalized = normalize(openApiElement);

          expect(toValue(normalized)).toMatchSnapshot();
        });
      });

      describe('given denormalized Security Requirements Objects', () => {
        test('should normalize', () => {
          const spec = globalThis.loadJsonFile(
            path.join(fixturesPath, 'security-requirements.json')
          );
          const openApiElement = OpenApi3_1Element.refract(spec);
          const normalized = normalize(openApiElement);

          expect(toValue(normalized)).toMatchSnapshot();
        });
      });

      describe('given denormalized Servers Objects', () => {
        test('should normalize', () => {
          const spec = globalThis.loadJsonFile(path.join(fixturesPath, 'servers.json'));
          const openApiElement = OpenApi3_1Element.refract(spec);
          const normalized = normalize(openApiElement);

          expect(toValue(normalized)).toMatchSnapshot();
        });
      });

      describe('given element non compatible with ObjectElement', () => {
        test('should skip normalization', () => {
          const element = new StringElement('test');
          const normalized = normalize(element);

          expect(normalized).toEqual(element);
        });
      });

      describe('given spec is already normalized', () => {
        test('should skip normalization', () => {
          const spec = globalThis.loadJsonFile(path.join(fixturesPath, 'servers.json'));
          const openApiElement = OpenApi3_1Element.refract(spec);
          const normalized = normalize(openApiElement);
          const doubleNormalized = normalize(openApiElement);

          expect(normalized).toEqual(doubleNormalized);
        });
      });
    });
  });
});
/* eslint-enable camelcase */
