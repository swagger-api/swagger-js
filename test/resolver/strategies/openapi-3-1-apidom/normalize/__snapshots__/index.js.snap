// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers normalize OpenAPI 3.1 given denormalized Header Object examples should normalize 1`] = `
{
  "$$normalized": true,
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "responses": {
          "200": {
            "headers": {
              "content-type": {
                "examples": {
                  "example1": {
                    "value": 2,
                  },
                },
                "schema": {
                  "examples": [
                    2,
                  ],
                  "type": "number",
                },
              },
            },
          },
        },
        "servers": [
          {
            "url": "/",
          },
        ],
      },
      "servers": [
        {
          "url": "/",
        },
      ],
    },
  },
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;

exports[`helpers normalize OpenAPI 3.1 given denormalized Operation.id fields should normalize 1`] = `
{
  "$$normalized": true,
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "__originalOperationId": "get operation ^",
        "operationId": "get_operation__",
        "servers": [
          {
            "url": "/",
          },
        ],
      },
      "servers": [
        {
          "url": "/",
        },
      ],
    },
  },
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;

exports[`helpers normalize OpenAPI 3.1 given denormalized Parameter Object examples should normalize 1`] = `
{
  "$$normalized": true,
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "parameters": [
        {
          "examples": {
            "example1": {
              "value": 2,
            },
          },
          "in": "query",
          "name": "param1",
          "schema": {
            "examples": [
              2,
            ],
            "type": "number",
          },
        },
      ],
      "servers": [
        {
          "url": "/",
        },
      ],
    },
  },
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;

exports[`helpers normalize OpenAPI 3.1 given denormalized Parameter Objects should normalize 1`] = `
{
  "$$normalized": true,
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "parameters": [
          {
            "in": "query",
            "name": "param1",
          },
          {
            "in": "query",
            "name": "param2",
          },
        ],
        "servers": [
          {
            "url": "/",
          },
        ],
      },
      "parameters": [
        {
          "in": "query",
          "name": "param1",
        },
        {
          "in": "query",
          "name": "param2",
        },
      ],
      "servers": [
        {
          "url": "/",
        },
      ],
    },
  },
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;

exports[`helpers normalize OpenAPI 3.1 given denormalized Security Requirements Objects should normalize 1`] = `
{
  "$$normalized": true,
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "security": [
          {
            "petstore_auth": [
              "write:pets",
              "read:pets",
            ],
          },
        ],
        "servers": [
          {
            "url": "/",
          },
        ],
      },
      "servers": [
        {
          "url": "/",
        },
      ],
    },
  },
  "security": [
    {
      "petstore_auth": [
        "write:pets",
        "read:pets",
      ],
    },
  ],
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;

exports[`helpers normalize OpenAPI 3.1 given denormalized Servers Objects should normalize 1`] = `
{
  "$$normalized": true,
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "servers": [
          {
            "description": "production server",
            "url": "https://example.com/",
          },
        ],
      },
      "servers": [
        {
          "description": "production server",
          "url": "https://example.com/",
        },
      ],
    },
  },
  "servers": [
    {
      "description": "production server",
      "url": "https://example.com/",
    },
  ],
}
`;
